sensorprocess/sensorprocess.go

```diff
@@ -12,39 +11,85 @@ import (
 	s "github.com/viamrobotics/viam-cartographer/sensors"
 )
 
-// defaultTime is used to check if timestamps have not been set yet.
-var defaultTime = time.Time{}
-
 // Config holds config needed throughout the process of adding a sensor reading to the cartofacade.
 type Config struct {
-	CartoFacade            cartofacade.Interface
-	sensorProcessStartTime time.Time
-	IsOnline               bool
-
-	Lidar            s.TimedLidar
-	currentLidarData *s.TimedLidarReadingResponse
+	CartoFacade cartofacade.Interface
+	IsOnline    bool
 
-	IMU            s.TimedMovementSensor
-	currentIMUData *s.TimedIMUReadingResponse
+	Lidar s.TimedLidar
+	IMU   s.TimedMovementSensor
 
 	Timeout                  time.Duration
 	InternalTimeout          time.Duration
 	Logger                   logging.Logger
 	RunFinalOptimizationFunc func(context.Context, time.Duration) error
-
-	Mutex *sync.Mutex
 }
 
-// Update currentIMUData under a mutex lock.
-func (config *Config) updateMutexProtectedIMUData(data s.TimedIMUReadingResponse) {
-	config.Mutex.Lock()
-	config.currentIMUData = &data
-	config.Mutex.Unlock()
+// StartOfflineSensorProcess starts the process of adding lidar and movement sensor data
+// in a deterministically defined order to cartographer.
+func (config *Config) StartOfflineSensorProcess(ctx context.Context) bool {
+	// get the initial lidar reading
+	lidarReading, lidarEndOfDataSetReached, err := getTimedLidarReading(ctx, config)
+	if err != nil || lidarEndOfDataSetReached {
+		return lidarEndOfDataSetReached
+	}
+
+	var imuReading s.TimedIMUReadingResponse
+	if config.IMU != nil {
+		// get the initial IMU reading; discard all IMU readings that were recorded before the first lidar reading
+		for {
+			msReading, msEndOfDataSetReached, err := getTimedMovementSensorReading(ctx, config)
+			if err != nil || msEndOfDataSetReached {
+				return msEndOfDataSetReached
+			}
+			imuReading = *msReading.TimedIMUResponse
+			if imuReading.ReadingTime.Equal(lidarReading.ReadingTime) ||
+				imuReading.ReadingTime.After(lidarReading.ReadingTime) {
+				break
+			}
+		}
+	}
+
+	// loop over all the data until one of the datasets has reached its end
+	for {
+		select {
+		case <-ctx.Done():
+			return false
+		default:
+			// insert the reading with the earliest time stamp
+			if config.IMU == nil ||
+				lidarReading.ReadingTime.Before(imuReading.ReadingTime) ||
+				lidarReading.ReadingTime.Equal(imuReading.ReadingTime) {
+				if err := config.tryAddLidarReadingUntilSuccess(ctx, lidarReading); err != nil {
+					return false
+				}
+				lidarReading, lidarEndOfDataSetReached, err = getTimedLidarReading(ctx, config)
+				if err != nil || lidarEndOfDataSetReached {
```

Comment:
same comment as above, why are we allowing non-EndOfDatsets errors to be handled as if they are? Are we saying that any and all error should be treated like the end of the dataset has been reached?